File: app/src/main/java/org/gnucash/android/ui/report/linechart/CashFlowLineChartFragment.java
/*
 * Copyright (c) 2015 Oleksandr Tyshkovets <olexandr.tyshkovets@gmail.com>
 * Copyright (c) 2015 Ngewi Fet <ngewif@gmail.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.gnucash.android.ui.report.linechart;

import static org.gnucash.android.util.ColorExtKt.parseColor;

import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.ColorInt;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.github.mikephil.charting.components.Legend;
import com.github.mikephil.charting.components.LimitLine;
import com.github.mikephil.charting.data.Entry;
import com.github.mikephil.charting.data.LineData;
import com.github.mikephil.charting.data.LineDataSet;
import com.github.mikephil.charting.formatter.LargeValueFormatter;
import com.github.mikephil.charting.highlight.Highlight;
import com.github.mikephil.charting.interfaces.datasets.ILineDataSet;

import org.gnucash.android.R;
import org.gnucash.android.databinding.FragmentLineChartBinding;
import org.gnucash.android.db.DatabaseSchema.AccountEntry;
import org.gnucash.android.model.Account;
import org.gnucash.android.model.AccountType;
import org.gnucash.android.model.Commodity;
import org.gnucash.android.model.Money;
import org.gnucash.android.model.Price;
import org.gnucash.android.ui.report.IntervalReportFragment;
import org.gnucash.android.ui.report.ReportType;
import org.gnucash.android.ui.report.ReportsActivity;
import org.gnucash.android.util.DateExtKt;
import org.joda.time.LocalDateTime;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import timber.log.Timber;

/**
 * Fragment for line chart reports
 *
 * @author Oleksandr Tyshkovets <olexandr.tyshkovets@gmail.com>
 * @author Ngewi Fet <ngewif@gmail.com>
 */
public class CashFlowLineChartFragment extends IntervalReportFragment {

    private static final int ANIMATION_DURATION = 3000;
    private static final int NO_DATA_BAR_COUNTS = 5;
    private static final int[] LINE_COLORS = {
        parseColor("#68F1AF"), parseColor("#cc1f09"), parseColor("#EE8600"),
        parseColor("#1469EB"), parseColor("#B304AD"),
    };
    private static final int[] FILL_COLORS = {
        parseColor("#008000"), parseColor("#FF0000"), parseColor("#BE6B00"),
        parseColor("#0065FF"), parseColor("#8F038A"),
    };

    private FragmentLineChartBinding mBinding;

    @Override
    public View inflateView(LayoutInflater inflater, ViewGroup container) {
        mBinding = FragmentLineChartBinding.inflate(inflater, container, false);
        return mBinding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        final Context context = view.getContext();

        @ColorInt int textColorPrimary = getTextColor(context);

        mBinding.lineChart.setOnChartValueSelectedListener(this);
        mBinding.lineChart.getXAxis().setDrawGridLines(false);
        mBinding.lineChart.getXAxis().setTextColor(textColorPrimary);
        mBinding.lineChart.getAxisRight().setEnabled(false);
        mBinding.lineChart.getAxisLeft().enableGridDashedLine(4.0f, 4.0f, 0);
        mBinding.lineChart.getAxisLeft().setValueFormatter(new LargeValueFormatter(mCommodity.getSymbol()));
        mBinding.lineChart.getAxisLeft().setTextColor(textColorPrimary);
        Legend legend = mBinding.lineChart.getLegend();
        legend.setTextColor(textColorPrimary);
    }

    @Override
    public ReportType getReportType() {
        return ReportType.LINE_CHART;
    }

    /**
     * Returns a data object that represents a user data of the specified account types
     *
     * @param accountTypes account's types which will be displayed
     * @return a {@code LineData} instance that represents a user data
     */
    @NonNull
    private LineData getData(@NonNull Context context, List<AccountType> accountTypes) {
        Timber.i("getData for %s", accountTypes);
        calculateEarliestAndLatestTimestamps(accountTypes);
        ReportsActivity.GroupInterval groupInterval = mGroupInterval;
        LocalDateTime startDate = mReportPeriodStart;
        LocalDateTime endDate = mReportPeriodEnd;

        List<ILineDataSet> dataSets = new ArrayList<>();
        for (AccountType accountType : accountTypes) {
            List<Entry> entries = getEntryList(accountType, groupInterval, startDate, endDate);
            LineDataSet dataSet = new LineDataSet(entries, getLabel(context, accountType));
            dataSet.setDrawFilled(true);
            dataSet.setLineWidth(2);
            dataSet.setColor(LINE_COLORS[dataSets.size()]);
            dataSet.setFillColor(FILL_COLORS[dataSets.size()]);

            dataSets.add(dataSet);
        }

        LineData lineData = new LineData(dataSets);
        if (getYValueSum(lineData) == 0) {
            isChartDataPresent = false;
            return getEmptyData(context);
        }
        lineData.setValueTextColor(getTextColor(context));
        return lineData;
    }

    /**
     * Returns a data object that represents situation when no user data available
     *
     * @return a {@code LineData} instance for situation when no user data available
     */
    private LineData getEmptyData(@NonNull Context context) {
        List<Entry> yValues = new ArrayList<>();
        boolean isEven = true;
        for (int i = 0; i < NO_DATA_BAR_COUNTS; i++) {
            yValues.add(new Entry(i, isEven ? 5f : 4.5f));
            isEven = !isEven;
        }
        LineDataSet dataSet = new LineDataSet(yValues, context.getString(R.string.label_chart_no_data));
        dataSet.setDrawFilled(true);
        dataSet.setDrawValues(false);
        dataSet.setColor(NO_DATA_COLOR);
        dataSet.setFillColor(NO_DATA_COLOR);

        return new LineData(dataSet);
    }

    /**
     * Returns entries which represent a user data of the specified account type
     *
     * @param accountType   account's type which user data will be processed
     * @param groupInterval
     * @return entries which represent a user data
     */
    private List<Entry> getEntryList(
        @NonNull AccountType accountType,
        @NonNull ReportsActivity.GroupInterval groupInterval,
        @Nullable LocalDateTime startEntries,
        @Nullable LocalDateTime endEntries
    ) {
        final Commodity commodity = mCommodity;
        List<Entry> entries = new ArrayList<>();

        LocalDateTime startDate = startEntries;
        if (startDate == null) {
            Long startTime = earliestTimestamps.get(accountType);
            if (startTime != null) {
                startDate = new LocalDateTime(startTime);
            } else {
                return entries;
            }
        }
        LocalDateTime endDate = endEntries;
        if (endDate == null) {
            Long endTime = latestTimestamps.get(accountType);
            if (endTime != null) {
                endDate = new LocalDateTime(endTime);
            } else {
                endDate = LocalDateTime.now();
            }
        }
        final LocalDateTime earliestDate = earliestTransactionTimestamp;
        int xAxisOffset = getDateDiff(groupInterval, earliestDate, startDate);
        int count = getDateDiff(groupInterval, startDate, endDate);
        LocalDateTime startPeriod = startDate;
        LocalDateTime endPeriod = endDate;
        switch (groupInterval) {
            case MONTH:
                endPeriod = startPeriod.plusMonths(1);
                break;
            case QUARTER:
                startPeriod = startPeriod.withMonthOfYear(DateExtKt.getFirstQuarterMonth(startPeriod)).dayOfMonth().withMinimumValue();
                endPeriod = startPeriod.plusMonths(3);
                break;
            case YEAR:
                endPeriod = startPeriod.plusYears(1);
                break;
        }

        String where = AccountEntry.COLUMN_TYPE + "=?"
            + " AND " + AccountEntry.COLUMN_PLACEHOLDER + " = 0"
            + " AND " + AccountEntry.COLUMN_TEMPLATE + " = 0";
        String[] whereArgs = new String[]{accountType.name()};
        List<Account> accounts = mAccountsDbAdapter.getSimpleAccounts(where, whereArgs, null);

        for (int i = 0, x = xAxisOffset; i < count; i++, x++) {
            long startTime = DateExtKt.toMillis(startPeriod);
            long endTime = DateExtKt.toMillis(endPeriod);
            Money balance = Money.createZeroInstance(commodity);
            Map<String, Money> balances = mAccountsDbAdapter.getAccountsBalances(accounts, startTime, endTime);
            for (Money accountBalance : balances.values()) {
                Price price = pricesDbAdapter.getPrice(accountBalance.getCommodity(), commodity);
                if (price == null) continue;
                accountBalance = accountBalance.times(price);
                balance = balance.plus(accountBalance);
            }
            Timber.d("%s %s %s - %s %s", accountType, groupInterval, startPeriod, endPeriod, balance);

            startPeriod = endPeriod;
            switch (groupInterval) {
                case MONTH:
                    endPeriod = endPeriod.plusMonths(1);
                    break;
                case QUARTER:
                    endPeriod = endPeriod.plusMonths(3);
                    break;
                case YEAR:
                    endPeriod = endPeriod.plusYears(1);
                    break;
            }

            if (balance.isAmountZero()) continue;
            float value = balance.toFloat();
            entries.add(new Entry(x, value));
        }

        return entries;
    }

    @Override
    public boolean requiresAccountTypeOptions() {
        return false;
    }

    @Override
    protected void generateReport(@NonNull Context context) {
        LineData lineData = getData(context, accountTypes);
        mBinding.lineChart.setData(lineData);
        isChartDataPresent = true;
    }

    @Override
    protected void displayReport() {
        if (!isChartDataPresent) {
            final Context context = mBinding.lineChart.getContext();
            mBinding.lineChart.getAxisLeft().setAxisMaxValue(10);
            mBinding.lineChart.getAxisLeft().setDrawLabels(false);
            mBinding.lineChart.getXAxis().setDrawLabels(false);
            mBinding.lineChart.setTouchEnabled(false);
            mSelectedValueTextView.setText(context.getString(R.string.label_chart_no_data));
        } else {
            mBinding.lineChart.animateX(ANIMATION_DURATION);
        }
        mBinding.lineChart.invalidate();
    }

    @Override
    public void onPrepareOptionsMenu(@NonNull Menu menu) {
        super.onPrepareOptionsMenu(menu);
        menu.findItem(R.id.menu_toggle_average_lines).setVisible(isChartDataPresent);
        showLegend(menu.findItem(R.id.menu_toggle_legend).isChecked());
        showAverageLines(menu.findItem(R.id.menu_toggle_average_lines).isChecked());
        // hide pie/bar chart specific menu items
        menu.findItem(R.id.menu_order_by_size).setVisible(false);
        menu.findItem(R.id.menu_toggle_labels).setVisible(false);
        menu.findItem(R.id.menu_percentage_mode).setVisible(false);
        menu.findItem(R.id.menu_group_other_slice).setVisible(false);
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        switch (item.getItemId()) {
            case R.id.menu_toggle_legend: {
                item.setChecked(!item.isChecked());
                showLegend(item.isChecked());
                return true;
            }

            case R.id.menu_toggle_average_lines:
                item.setChecked(!item.isChecked());
                showAverageLines(item.isChecked());
                return true;

            default:
                return super.onOptionsItemSelected(item);
        }
    }

    @Override
    public void onValueSelected(Entry e, Highlight h) {
        if (e == null) return;
        float value = e.getY();
        int dataSetIndex = h.getDataSetIndex();
        LineData data = mBinding.lineChart.getData();
        ILineDataSet dataSet = data.getDataSetByIndex(dataSetIndex);
        if (dataSet == null) return;
        String label = dataSet.getLabel();
        float total = getYValueSum(dataSet);
        float percent = (total != 0f) ? ((value * 100) / total) : 0f;
        mSelectedValueTextView.setText(formatSelectedValue(label, value, percent));
    }

    private void showLegend(boolean isVisible) {
        mBinding.lineChart.getLegend().setEnabled(isVisible);
        mBinding.lineChart.invalidate();
    }

    private void showAverageLines(boolean isVisible) {
        mBinding.lineChart.getAxisLeft().removeAllLimitLines();
        if (isVisible) {
            for (ILineDataSet dataSet : mBinding.lineChart.getData().getDataSets()) {
                int entryCount = dataSet.getEntryCount();
                float limit = 0f;
                if (entryCount > 0) {
                    limit = dataSet.getYMin() + (getYValueSum(dataSet) / entryCount);
                }
                LimitLine line = new LimitLine(limit, dataSet.getLabel());
                line.enableDashedLine(10, 5, 0);
                line.setLineColor(dataSet.getColor());
                mBinding.lineChart.getAxisLeft().addLimitLine(line);
            }
        }
        mBinding.lineChart.invalidate();
    }

}

File: app/src/main/java/org/gnucash/android/ui/report/linechart/CashFlowLineChartFragment.kt
/*
 * Copyright (c) 2015 Oleksandr Tyshkovets <olexandr.tyshkovets@gmail.com>
 * Copyright (c) 2015 Ngewi Fet <ngewif@gmail.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.gnucash.android.ui.report.linechart

import android.content.Context
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import androidx.annotation.ColorInt
import com.github.mikephil.charting.charts.LineChart
import com.github.mikephil.charting.components.LimitLine
import com.github.mikephil.charting.data.Entry
import com.github.mikephil.charting.data.LineData
import com.github.mikephil.charting.data.LineDataSet
import com.github.mikephil.charting.formatter.LargeValueFormatter
import com.github.mikephil.charting.highlight.Highlight
import com.github.mikephil.charting.interfaces.datasets.ILineDataSet
import org.gnucash.android.R
import org.gnucash.android.databinding.FragmentChartBinding
import org.gnucash.android.db.DatabaseSchema.AccountEntry
import org.gnucash.android.model.AccountType
import org.gnucash.android.model.Money.Companion.createZeroInstance
import org.gnucash.android.ui.report.IntervalReportFragment
import org.gnucash.android.ui.report.ReportType
import org.gnucash.android.ui.report.ReportsActivity.GroupInterval
import org.gnucash.android.util.getFirstQuarterMonth
import org.gnucash.android.util.parseColor
import org.gnucash.android.util.toMillis
import org.joda.time.LocalDateTime
import timber.log.Timber

/**
 * Fragment for line chart reports
 *
 * @author Oleksandr Tyshkovets <olexandr.tyshkovets@gmail.com>
 * @author Ngewi Fet <ngewif@gmail.com>
 */
class CashFlowLineChartFragment : IntervalReportFragment<LineData>() {
    private var binding: FragmentChartBinding? = null
    private var chart: LineChart? = null

    override fun inflateView(inflater: LayoutInflater, container: ViewGroup?): View {
        val binding = FragmentChartBinding.inflate(inflater, container, false)
        this.binding = binding
        this.selectedValueTextView = binding.selectedChartSlice
        return binding.root
    }

    override val reportType: ReportType = ReportType.LINE_CHART

    /**
     * Returns a data object that represents a user data of the specified account types
     *
     * @param accountTypes account's types which will be displayed
     * @return a `LineData` instance that represents a user data
     */
    private fun getData(context: Context, accountTypes: List<AccountType>): LineData {
        Timber.i("getData for %s", accountTypes)
        calculateEarliestAndLatestTimestamps(accountTypes)
        val groupInterval = this.groupInterval
        val startDate = reportPeriodStart
        val endDate = reportPeriodEnd

        val dataSets = mutableListOf<ILineDataSet>()
        for (accountType in accountTypes) {
            val index = dataSets.size
            val entries = getEntryList(accountType, groupInterval, startDate, endDate)
            val dataSet = LineDataSet(entries, getLabel(context, accountType))
            dataSet.setDrawFilled(true)
            dataSet.lineWidth = 2f
            dataSet.color = LINE_COLORS[index]
            dataSet.fillColor = FILL_COLORS[index]
            dataSets.add(dataSet)
        }

        val lineData = LineData(dataSets)
        lineData.setValueTextColor(getTextColor(context))
        return lineData
    }

    /**
     * Returns a data object that represents situation when no user data available
     *
     * @return a `LineData` instance for situation when no user data available
     */
    private fun getEmptyData(context: Context): LineData {
        val yValues = mutableListOf<Entry>()
        var isEven = true
        for (i in 0 until NO_DATA_BAR_COUNTS) {
            yValues.add(Entry(i.toFloat(), if (isEven) 5f else 4.5f))
            isEven = !isEven
        }
        val dataSet = LineDataSet(yValues, context.getString(R.string.label_chart_no_data))
        dataSet.setDrawFilled(true)
        dataSet.setDrawValues(false)
        dataSet.color = NO_DATA_COLOR
        dataSet.fillColor = NO_DATA_COLOR

        return LineData(dataSet)
    }

    private fun isEmpty(data: LineData): Boolean {
        return (data.dataSetCount == 0) ||
                (data.entryCount == 0) ||
                ((data.yMin <= DATA_EMPTY) && (data.yMax <= DATA_EMPTY))
    }

    /**
     * Returns entries which represent a user data of the specified account type
     *
     * @param accountType   account's type which user data will be processed
     * @param groupInterval
     * @return entries which represent a user data
     */
    private fun getEntryList(
        accountType: AccountType,
        groupInterval: GroupInterval,
        startEntries: LocalDateTime?,
        endEntries: LocalDateTime?
    ): List<Entry> {
        val commodity = this.commodity
        val entries = mutableListOf<Entry>()

        var startDate = startEntries
        if (startDate == null) {
            val startTime = earliestTimestamps[accountType]
            if (startTime != null) {
                startDate = LocalDateTime(startTime)
            } else {
                return entries
            }
        }
        var endDate = endEntries
        if (endDate == null) {
            val endTime = latestTimestamps[accountType]
            endDate = if (endTime != null) {
                LocalDateTime(endTime)
            } else {
                LocalDateTime.now()
            }
        }
        val earliestDate = earliestTransactionTimestamp!!
        val xAxisOffset = getDateDiff(groupInterval, earliestDate, startDate)
        val count = getDateDiff(groupInterval, startDate, endDate)
        var startPeriod: LocalDateTime = startDate
        var endPeriod = endDate!!
        when (groupInterval) {
            GroupInterval.MONTH -> endPeriod = startPeriod.plusMonths(1)
            GroupInterval.QUARTER -> {
                startPeriod = startPeriod.withMonthOfYear(startPeriod.getFirstQuarterMonth())
                    .dayOfMonth()
                    .withMinimumValue()
                endPeriod = startPeriod.plusMonths(3)
            }

            GroupInterval.YEAR -> endPeriod = startPeriod.plusYears(1)
            else -> Unit
        }

        val where = (AccountEntry.COLUMN_TYPE + "=?"
                + " AND " + AccountEntry.COLUMN_PLACEHOLDER + " = 0"
                + " AND " + AccountEntry.COLUMN_TEMPLATE + " = 0")
        val whereArgs = arrayOf<String?>(accountType.name)
        val accounts = accountsDbAdapter.getAllRecords(where, whereArgs)

        var i = 0
        var x = xAxisOffset
        while (i < count) {
            val startTime = startPeriod.toMillis()
            val endTime = endPeriod.toMillis()
            var balance = createZeroInstance(commodity)
            val balances = accountsDbAdapter.getAccountsBalances(accounts, startTime, endTime)
            for (accountBalance in balances.values) {
                var accountBalance = accountBalance
                val price =
                    pricesDbAdapter.getPrice(accountBalance.commodity, commodity) ?: continue
                accountBalance *= price
                balance += accountBalance
            }
            Timber.d(
                "%s %s %s - %s %s",
                accountType,
                groupInterval,
                startPeriod,
                endPeriod,
                balance
            )

            startPeriod = endPeriod
            when (groupInterval) {
                GroupInterval.MONTH -> endPeriod = endPeriod.plusMonths(1)
                GroupInterval.QUARTER -> endPeriod = endPeriod.plusMonths(3)
                GroupInterval.YEAR -> endPeriod = endPeriod.plusYears(1)
                else -> Unit
            }

            if (balance.isAmountZero) {
                i++
                x++
                continue
            }
            val value = balance.toFloat()
            entries.add(Entry(x.toFloat(), value))
            i++
            x++
        }

        return entries
    }

    override fun requiresAccountTypeOptions(): Boolean {
        return false
    }

    override fun generateReport(context: Context): LineData {
        isChartDataPresent = false
        val data = getData(context, accountTypes)
        if (isEmpty(data)) {
            return getEmptyData(context)
        }
        isChartDataPresent = true
        return data
    }

    override fun displayReport(data: LineData) {
        val binding = binding ?: return
        val context = binding.root.context
        val isChartDataPresent = isChartDataPresent
        val selectedValueTextView = binding.selectedChartSlice
        @ColorInt val textColorPrimary = getTextColor(context)

        val chart = LineChart(context).apply {
            id = R.id.chart
            setOnChartValueSelectedListener(this@CashFlowLineChartFragment)
            xAxis.setDrawGridLines(false)
            xAxis.textColor = textColorPrimary
            axisRight.isEnabled = false
            axisLeft.enableGridDashedLine(4.0f, 4.0f, 0f)
            axisLeft.valueFormatter = LargeValueFormatter(commodity.symbol)
            axisLeft.textColor = textColorPrimary
            legend.textColor = textColorPrimary

            this.data = data

            if (isChartDataPresent) {
                selectedValueTextView.text = null
                animateX(ANIMATION_DURATION)
            } else {
                selectedValueTextView.setText(R.string.label_chart_no_data)
                axisLeft.setAxisMaxValue(10f)
                axisLeft.setDrawLabels(false)
                xAxis.setDrawLabels(false)
                setTouchEnabled(false)
                clearAnimation()
            }
            highlightValues(null)
        }
        this.chart = chart

        binding.chartContainer.apply {
            removeAllViews()
            addView(
                chart,
                ViewGroup.LayoutParams(
                    ViewGroup.LayoutParams.MATCH_PARENT,
                    ViewGroup.LayoutParams.MATCH_PARENT
                )
            )
        }
    }

    override fun onPrepareOptionsMenu(menu: Menu) {
        super.onPrepareOptionsMenu(menu)
        menu.findItem(R.id.menu_toggle_average_lines).isVisible = isChartDataPresent
        showLegend(menu.findItem(R.id.menu_toggle_legend).isChecked)
        showAverageLines(menu.findItem(R.id.menu_toggle_average_lines).isChecked)
        // hide pie/bar chart specific menu items
        menu.findItem(R.id.menu_order_by_size).isVisible = false
        menu.findItem(R.id.menu_toggle_labels).isVisible = false
        menu.findItem(R.id.menu_percentage_mode).isVisible = false
        menu.findItem(R.id.menu_group_other_slice).isVisible = false
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.menu_toggle_legend -> {
                item.isChecked = !item.isChecked
                showLegend(item.isChecked)
                return true
            }

            R.id.menu_toggle_average_lines -> {
                item.isChecked = !item.isChecked
                showAverageLines(item.isChecked)
                return true
            }

            else -> return super.onOptionsItemSelected(item)
        }
    }

    override fun onValueSelected(e: Entry?, h: Highlight) {
        val chart = chart ?: return
        if (e == null) return
        val value = e.y
        val dataSetIndex = h.dataSetIndex
        val data = chart.data
        val dataSet = data.getDataSetByIndex(dataSetIndex) ?: return
        val label = dataSet.label ?: return
        val total = getYValueSum<Entry>(dataSet)
        val percent = if (total != 0f) ((value * 100) / total) else 0f
        selectedValueTextView?.text = formatSelectedValue(label, value, percent)
    }

    private fun showLegend(isVisible: Boolean) {
        val lineChart = chart ?: return
        lineChart.legend.isEnabled = isVisible
        lineChart.invalidate()
    }

    private fun showAverageLines(isVisible: Boolean) {
        val lineChart = chart ?: return
        lineChart.axisLeft.removeAllLimitLines()
        if (isVisible) {
            for (dataSet in lineChart.data.dataSets) {
                val entryCount = dataSet.entryCount
                var limit = 0f
                if (entryCount > 0) {
                    limit = dataSet.yMin + (getYValueSum<Entry>(dataSet) / entryCount)
                }
                val line = LimitLine(limit, dataSet.label)
                line.enableDashedLine(10f, 5f, 0f)
                line.lineColor = dataSet.color
                lineChart.axisLeft.addLimitLine(line)
            }
        }
        lineChart.invalidate()
    }

    companion object {
        private const val ANIMATION_DURATION = 3000
        private const val NO_DATA_BAR_COUNTS = 5
        private val LINE_COLORS = intArrayOf(
            parseColor("#68F1AF")!!, parseColor("#cc1f09")!!, parseColor("#EE8600")!!,
            parseColor("#1469EB")!!, parseColor("#B304AD")!!,
        )
        private val FILL_COLORS = intArrayOf(
            parseColor("#008000")!!, parseColor("#FF0000")!!, parseColor("#BE6B00")!!,
            parseColor("#0065FF")!!, parseColor("#8F038A")!!,
        )
    }
}
